Show "bip125-replaceable" flag, when retrieving mempool entries

dexX7 opened the pull request:
This pull request adds a flag "bip125-replaceable" to the mempool RPCs getrawmempool, getmempoolentry, getmempoolancestors and getmempooldescendants, which indicates whether an unconfirmed transaction might be replaced.

Initially the flag was added to the raw transaction RPCs, but thanks to @conscott, it was moved to the mempool RPCs, which actually have access to the mempool.

~~This pull request adds a flag "bip125-replaceable" to the RPCs "getrawtransaction" and "decoderawtransaction", which indicates, whether a transaction signals BIP 125 replaceability.~~

There was some discussion in #7817, whether showing replaceability in the UI could lead to the false assumption that transactions that don't signal BIP 125 are truely non-replaceable, but given that this PR tackles the raw transaction interface, which is a rather low level tool, I believe having this extra piece of information isn't bad.

dexX7 commented:
Please note, I copied SignalsOptInRBF() from policy/rbf.cpp into core_write.cpp.

This is probably not a good way to, but was done, because rbf.cpp isn't avaliable for LIBBITCOIN_COMMON. However, simply moving rbf.cpp from LIBBITCOIN_SERVER to LIBBITCOIN_COMMON doesn't work, because IsRBFOptIn() in rbf.cpp has mempool dependencies, which are also not available.

As alternative I could imagine to split rbf.cpp into rbf.cpp and rbf_mempool.cpp?

MarcoFalke commented on src/core_write.cpp:
```diff
@@ -162,6 +172,7 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
     entry.pushKV("size", (int)::GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION));
     entry.pushKV("vsize", (GetTransactionWeight(tx) + WITNESS_SCALE_FACTOR - 1) / WITNESS_SCALE_FACTOR);
     entry.pushKV("locktime", (int64_t)tx.nLockTime);
+    entry.pushKV("replaceable", SignalsOptInRBF(tx));
```
Imo this should be the same key-value as the `bip125-replaceable` in the rpcwallet

dexX7 replied:
Good idea! I updated the PR.

MarcoFalke commented on src/core_write.cpp:
```diff
@@ -162,6 +172,7 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
     entry.pushKV("size", (int)::GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION));
     entry.pushKV("vsize", (GetTransactionWeight(tx) + WITNESS_SCALE_FACTOR - 1) / WITNESS_SCALE_FACTOR);
     entry.pushKV("locktime", (int64_t)tx.nLockTime);
+    entry.pushKV("bip125-replaceable", SignalsOptInRBF(tx));
```
`SignalsOptInRBF` is *not* bip125-replacable.

* Confirmed transaction can never be replaced in the mempool
* Unconfirmed transactions that depend on a mempool-transaction that signals opt in rbf can be replaced

dexX7 replied:
Would you suggest to rename the flag to something like `signals-rbf`?

laanwj replied:
If we do this I'd say rename it to `signals-rbf` or such, which is also the name of the function already. And not make the raw transactions utility function decoding functionality depend on the mempool.

MarcoFalke replied:
In which case I tend to Concept NACK. Just returning if the flag is set contradicts the bip125-replacability rules.

conscott replied:
Sorry to hijack.

I had considered adding a flag like this to the verbose output of `getrawmempool`. Assuming this idea were supported, based on your feedback @MarcoFalke - is it fair to say that you think it makes more sense to list a mempool transaction as `replaceable` (covering both explicit signaling and implicit based on ancestors) rather than just `signals-rbf` (just the explicit case)? 

Do you think it would be appropriate to list that in `getrawmempool`?

dexX7 committed:
Add "bip125-replaceable" flag to mempool RPCs

This affects getrawmempool, getmempoolentry, getmempoolancestors and getmempooldescendants.

dexX7 committed:
Update functional RBF test to check replaceable flag

dexX7 commented:
@conscott brought up the idea of adding the flag to mempool entries and mempool RPCs. I really like the idea, because these calls are actually mempool aware and provide full access to the mempool and thus also allow to check, whether a transaction is really replaceable instead of simply signaling replaceablity.

I updated the PR accordingly, because I believe this is the way to go. @conscott if you want to submit this as PR, feel free to do so though, and I'll close this one.

instagibbs commented on src/rpc/blockchain.cpp:
```diff
@@ -418,6 +420,16 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
+    // Add opt-in RBF status
+    std::string rbfStatus = "no";
+    RBFTransactionState rbfState = IsRBFOptIn(tx, mempool);
+    if (rbfState == RBF_TRANSACTIONSTATE_UNKNOWN)
```
please use brackets for multi-line conditionals

dexX7 replied:
Thanks, updated.

instagibbs reviewed (approved):
utACK, minus style nits https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md

Sjors commented on test/functional/feature_rbf.py:
```diff
@@ -427,6 +427,9 @@ def test_opt_in(self):
         tx1a_hex = txToHex(tx1a)
         tx1a_txid = self.nodes[0].sendrawtransaction(tx1a_hex, True)
 
+        # This transaction isn't shown as replaceable
```
Can you add a test for `unknown`?

instagibbs commented:
re-ACK

Sjors reviewed (approved):
tACK eef1b43

dexX7 commented on test/functional/feature_rbf.py:
```diff
@@ -427,6 +427,9 @@ def test_opt_in(self):
         tx1a_hex = txToHex(tx1a)
         tx1a_txid = self.nodes[0].sendrawtransaction(tx1a_hex, True)
 
+        # This transaction isn't shown as replaceable
```
Hmm.. this is going to be a bit difficult: the `"unknown"` case is only triggered, if the transaction isn't found in the mempool.

In the case of the RPCs this path might even get removed, given that we iterate over mempool entries, which are indeed in the mempool and thus never `"unknown"`.

Sjors replied:
Such a test might require mocking `IsRBFOptIn` in `blockchain.cpp`, but that seems overkill here.

conscott commented on src/rpc/blockchain.cpp:
```diff
@@ -418,6 +420,17 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
+    std::string rbfStatus = "no";
+    RBFTransactionState rbfState = IsRBFOptIn(tx, mempool);
+    if (rbfState == RBF_TRANSACTIONSTATE_UNKNOWN) {
+        rbfStatus = "unknown";
```
I think this can just be an `assert(bfState != RBF_TRANSACTIONSTATE_UNKNOWN)`. Looking at [rbf.cpp](https://github.com/bitcoin/bitcoin/blob/master/src/policy/rbf.cpp#L27-L32), the unknown state should only occur if the transaction in question is not found in the mempool, and this function only processes transactions that are currently in the mempool (with mempool lock held). 

If the above is correct (I believe it is), the `bip125-replaceable` field can just become a boolean.

Sjors replied:
I'm not a fan of an assert here. It if it's really not possible, it's better to put the assert in `bf.cpp` and remove this state from the enum.

Sjors replied:
Actually: what about a situation where a descendant is missing from the mempool (because it somehow got evicted)?

conscott replied:
I should have been more clear. In general it is possible to return `RBF_TRANSACTIONSTATE_UNKNOWN` from `IsRBFOptIn()`, however this call stack in particular (all mempool related rpc's) only calls `IsRBFOptIn()` on transactions we know to be in the mempool. So, the assert would belong here, not in `rfb.cpp`. I understand this still may not be the best option, I just wanted to make my language clear since my comment above was a bit ambiguous. 

I am not sure I understand your situation where 'a descendant is missing from the mempool'. Can you elaborate please?

dexX7 replied:
I'm not a fan of using `assert` in non-consensus critical code, given it's massive impact, if it's hit (e.g. by accident in the future etc.).

In the case of the RPCs `"unknown"` isn't triggered, because they iterate over mempool entries, which are inherently part of the mempool, but I kind of tend to leave it there for the potential case `entryToJSON()` is called in a different context. I'd be fine with removing this path though, too.

> Actually: what about a situation where a descendant is missing from the mempool (because it somehow got evicted)?

Not an issue as far as I can see. We only care about ancestors and whether an unconfirmed ancestor signals RBF, which makes the chain replaceable.

conscott replied:
Agree about the 'descendant is missing from the mempool' not being an issue, or relevant, as far as I can see. 

I also see your point that in general the function could be called in a context where the tx is no longer in the mempool, but in cases strictly calling down from a mempool rpc, it seems like we should always 'know' rbf state in that context. In that sense, it may not make sense for the rpc itself to list `unknown` as an option - unless there actually is a case I am missing where we would not know the rbf state of a tx in our own mempool. 

Having said that, whether or not an `assert` is the right thing to do, I am also an open about. 

I will happily wait for someone more qualified than myself to chime in with their opinion :)

Sjors replied:
> this call stack in particular (all mempool related rpc's) only calls IsRBFOptIn() on transactions we know to be in the mempool

There's no guarantee all ancestors are in the mempool (someone might have deleted mempool.dat and it's still in the process of syncing, unless we guarantee order there). If a transaction in the mempool doesn't signal RBF, but its ancestors do, then it's replaceable. So `IsRBFOptIn` would return UNKNOWN in that case, I think..

sipa replied:
Mempool.dat is sorted in dependency order, so ancestors always appear earlier. Further, the mempool is at all times internally consistent (when observable): all mempool transactions either spend a transaction output from the chainstate (so from a confirmed transaction), or another mempool output. There are also never any double spends in the mempool.

conscott replied:
Any opinion on listing `unknown` state @MarcoFalke ?

MarcoFalke replied:
I don't have a strong opinion if the value of 'bip125-replacable' should be string (yes/no/unknown) or bool.

It seems `entryToJSON` can only be called when the entry is in the mempool (c.f. `mempool.vTxHashes[e.vTxHashesIdx].first.ToString()`). Thus, we can surely determine if it is replacable.

Instead of adding an `assert`, I'd prefer to throw some json exception that mentions that it never can be thrown.

I guess you could throw it in `entryToJSON` to make sure it is passed up to all potential callers?

dexX7 replied:
But does entryToJSON really care, whether it's in the mempool or not? I think this limits it's functionality in some way.

conscott replied:
`entryToJSON` specifically processes mempool transactions (`CTxMemPoolEntry &e`) - and I don't believe you should be able to have a CTxMemPoolEntry reference for something not currently in the mempool (it should guarantees this consistency), so in that sense, `entryToJson` should never be called in a "different context" (where the tx is not in mempool). 

Anyhow, I am not trying to get pedantic on this point. I want this change (and requested it) and I am okay with the current implementation.

conscott commented:
Concept ACK. Thanks for updating @dexX7 - just left a comment about the unknown state. Will test this out tomorrow.

dexX7 commented:
Regarding checking the `"unknown"` case: in the case of the RPCs this isn't triggered, because they iterate over mempool entries, which are inherently part of the mempool, but I kind of tend to leave it there for the potential case `entryToJSON()` is called in a different context. I'd be fine with removing this path though, too.

MeshCollider commented:
Concept ACK

conscott reviewed (commented):
Test ACK eef1b43b5860122bd7dc9270b5b5ba3f64b1dc0e

dexX7 commented:
Hey guys, just to clarify: given there were mixed reactions, do you insist on removing the branch for the "unknown" case and replace it with an exception or is this PR good to go as it is?

MarcoFalke commented:
I just fail to see why it makes sense to burden rpc users with a third value that is never actually set.

dexX7 commented:
Alright, updated. :)

dexX7 commented on src/rpc/blockchain.cpp:
```diff
@@ -375,7 +376,8 @@ std::string EntryDescriptionString()
            "        \"transactionid\",    (string) child transaction id\n"
-           "       ... ]\n";
+           "       ... ]\n"
+           "    \"bip125-replaceable\": \"yes|no\",  (string) Whether this transaction could be replaced due to BIP125 (replace-by-fee)\n";
```
Very good point.

dexX7 replied:
Updated.

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -419,6 +421,17 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
+    bool rbfStatus = false;
+    RBFTransactionState rbfState = IsRBFOptIn(tx, mempool);
+    if (rbfState == RBFTransactionState::UNKNOWN) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Transaction is not in mempool");
```
Throwing here seems nasty. Just don't add a key?

Eg: 

```C++
    switch (IsRBFOptIn(tx, mempool)) {
        case RBFTransactionState::FINAL:
            info.push_back(Pair("bip125-replaceable", false));
            break;
        case RBFTransactionState::REPLACEABLE_BIP125:
            info.push_back(Pair("bip125-replaceable", true));
            break;
        case RBFTransactionState::UNKNOWN:
            break;
    }
```

(Doing it this way also gets a compile-time warning if a new state is added later)

sdaftuar replied:
I think it'd be nice to have this produce the same output as the wallet rpc calls that provide the same information (which appears to be "yes"/"no"/"unknown").

MarcoFalke replied:
Agree with @achow101

luke-jr replied:
Agree with @sdaftuar

achow101 commented:
utACK 870bd4c73ddf494dc23c658bf0fb672ee0109158

MarcoFalke commented:
utACK 870bd4c

laanwj commented:
This seems mergeable, should do so after 0.17 branch as this is a feature
